packages/expect/src/__tests__/customEqualityTesters.test.ts

```diff
@@ -0,0 +1,245 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import type {Tester} from '@jest/expect-utils';
+import jestExpect from '../';
+
+const specialObjPropName = '$$special';
+const specialObjSymbol = Symbol('special test object type');
+
+interface SpecialObject {
+  $$special: symbol;
+  value: string;
+}
+
+function createSpecialObject(value: string) {
+  return {
+    [specialObjPropName]: specialObjSymbol,
+    value,
+  };
+}
+
+function isSpecialObject(a: unknown): a is SpecialObject {
+  return (
+    a != null &&
+    typeof a === 'object' &&
+    specialObjPropName in a &&
+    (a as any)[specialObjPropName] === specialObjSymbol
+  );
+}
+
+const specialObjTester: Tester = (
+  a: unknown,
+  b: unknown,
+): boolean | undefined => {
+  const isASpecial = isSpecialObject(a);
+  const isBSpecial = isSpecialObject(b);
+
+  if (isASpecial && isBSpecial) {
+    return true;
+  } else if ((isASpecial && !isBSpecial) || (!isASpecial && isBSpecial)) {
+    return false;
+  } else {
+    return undefined;
+  }
+};
+
+function* toIterator<T>(array: Array<T>): Iterator<T> {
+  for (const obj of array) {
+    yield obj;
+  }
+}
+
+declare module '../types' {
+  interface Matchers<R> {
+    toSpecialObjectEqual(expected: SpecialObject): R;
+  }
+}
+
+jestExpect.extend({
+  toSpecialObjectEqual(expected: SpecialObject, actual: SpecialObject) {
+    const result = this.equals(
+      expected,
+      actual,
+      jestExpect.customEqualityTesters,
```

Comment:
not sure whether ... having to / having the option to... specify override of the customEqualityTesters here is a good thing

Reply:
The idea here is to enable custom matchers to have the same capabilities as the built in matchers. Built-in matchers need to pass in custom equality testers to `equals` and so I would like custom matchers to have ability to do the same.

We could look overriding the `equals` implementation on `MatcherContext` to automatically pass in custom testers, but then a matcher couldn't customize the testers, which it might want to do. So my thinking was to just require custom matchers to do what the built-in matchers and pass in custom equality testers.

If we change the expect API to something like `expect.addEqualityTesters`, it would be nice to provide some API for custom matchers to access the added testers. Perhaps as a property on `MatcherContext`?